modules/spring-data-2.0-ext/src/main/java/org/apache/ignite/springdata20/repository/support/IgniteResourceProviderImpl.java

```diff
@@ -0,0 +1,171 @@
+import org.springframework.context.expression.StandardBeanExpressionResolver;
+import org.springframework.util.Assert;
+
+/** Represents default implementation of {@link IgniteResourceProvider} */
+public class IgniteResourceProviderImpl implements IgniteResourceProvider, ApplicationContextAware, DisposableBean {
+    /** Spring application expression resolver. */
+    private final BeanExpressionResolver expressionResolver = new StandardBeanExpressionResolver();
+
+    /** Ignite proxies associated with repositories. */
+    private final Map<Class<?>, IgniteProxy> igniteProxies = new ConcurrentHashMap<>();
+
+    /** Spring application context. */
+    private ApplicationContext ctx;
+
+    /** Spring application bean expression context. */
+    private BeanExpressionContext beanExpressionCtx;
+
+    /** {@inheritDoc} */
+    @Override public IgniteProxy igniteProxy(Class<?> repoInterface) {
+        return igniteProxies.computeIfAbsent(repoInterface, k -> createIgniteProxy(repoInterface));
+    }
+
+    /** {@inheritDoc} */
+    @Override public void setApplicationContext(ApplicationContext ctx) throws BeansException {
+        this.ctx = ctx;
+
+        beanExpressionCtx = new BeanExpressionContext(
+            new DefaultListableBeanFactory(ctx.getAutowireCapableBeanFactory()),
+            null);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void destroy() throws Exception {
+        Set<IgniteProxy> proxies = new HashSet<>(igniteProxies.values());
+
+        for (IgniteProxy proxy : proxies) {
+            if (proxy instanceof AutoCloseable)
+                ((AutoCloseable)proxy).close();
+        }
+    }
+
+    /**
+     * Creates {@link IgniteProxy} to be used for providing access to the Ignite cluster for specified repository.
+     *
+     * @param repoInterface {@link Class} instance of the repository interface.
+     * @return Instance of {@link IgniteProxy} associated with specified repository.
+     *
+     * @see RepositoryConfig
+     */
+    private IgniteProxy createIgniteProxy(Class<?> repoInterface) {
+        RepositoryConfig repoCfg = getRepositoryConfiguration(repoInterface);
+
+        try {
+            Object igniteInstanceBean = ctx.getBean(evaluateExpression(repoCfg.igniteInstance()));
+
+            if (igniteInstanceBean instanceof Ignite)
+                return new IgniteProxyImpl((Ignite)igniteInstanceBean);
+            else if (igniteInstanceBean instanceof IgniteClient)
+                return new IgniteClientProxy((IgniteClient)igniteInstanceBean);
+
+            throw new IllegalArgumentException("Invalid repository configuration [name=" + repoInterface.getName() +
+                "]. The Spring Bean corresponding to the \"igniteInstance\" property of repository configuration must" +
+                " be one of the following types: [" + Ignite.class.getName() + ", " + IgniteClient.class.getName() +
+                ']');
+        }
+        catch (BeansException ex) {
```

Comment:
I'm not sure about spring best practices, but usage of ctx.containsBean() looks better than build logic on exception handling. WDYT?

Reply:
In this case, I keep the logic as it is. Just moved it from IgniteRepository Factory to avoid unrelated changes. I think It can be done in separate ticket.